### PR TITLE
[Merged by Bors] - feat(RingTheory/AlgebraicIndependent): add `AlgebraicIndependent.aevalEquivField` and other results

### PR DESCRIPTION
**Pull request by acmepjz**
This PR mainly adds

- `AlgebraicIndependent.aevalEquivField`: canonical isomorphism between rational function field and the intermediate field generated by algebraically independent elements.
- `AlgebraicIndependent.reprField`: canonical map from the intermediate field generated by an algebraic independent family into the rational function field.

Other results:

- `AlgebraicIndependent.aeval_of_algebraicIndependent`: if `x = {x_i : A | i : ι}` and `f = {f_i : MvPolynomial ι R | i : ι}` are algebraically independent over `R`, then `{f_i(x) | i : ι}` is also algebraically independent over `R`.
- `AlgebraicIndependent.of_aeval`: conversely, if `{f_i(x) | i : ι}` is algebraically independent over `R`, then `{f_i : MvPolynomial ι R | i : ι}` is also algebraically independent over `R`.
- `AlgebraicIndependent.polynomial_aeval_of_transcendental`: if `{x_i : A | i : ι}` is algebraically independent over `R`, and for each `i`, `f_i : R[X]` is transcendental over `R`, then `{f_i(x_i) | i : ι}` is also algebraically independent over `R`.
- `AlgebraicIndependent.isEmpty_of_isAlgebraic`: if `A/R` is algebraic, then all algebraically independent families are empty.
- `IsTranscendenceBasis.isEmpty_iff_isAlgebraic`: if `x` is a transcendence basis of `A/R`, then it is empty if and only if `A/R` is algebraic.
- `IsTranscendenceBasis.nonempty_iff_transcendental`: if `x` is a transcendence basis of `A/R`, then it is not empty if and only if `A/R` is transcendental.
- `IsTranscendenceBasis.isAlgebraic_field`: if `x` is a transcendence basis of `E / F`, then the field extension `E / F(x)` is algebraic.

Also add `[of_]ringHom_of_comp_eq` for `[Algebra.]Transcendental` and `AlgebraicIndependent`.

Also add some results for `Subalgebra` inspired by `IntermediateField.isAlgebraic_iff`. Some of them are not back-ported to `IntermediateField` yet; may be for another PR if they are useful.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

- [x] depends on: #18661
- [x] depends on: #18984

### CODE REVIEW

File: Mathlib/RingTheory/AlgebraicIndependent.lean [hunk 1]
```diff
@@ -566,6 +704,40 @@ theorem IsTranscendenceBasis.isAlgebraic [Nontrivial R] (hx : IsTranscendenceBas
...
+    IsEmpty ι ↔ Algebra.IsAlgebraic R A := by
+  refine ⟨fun _ ↦ ?_, fun _ ↦ hx.1.isEmpty_of_isAlgebraic⟩
+  have := hx.isAlgebraic
+  rw [Set.range_eq_empty x, adjoin_empty] at this
+  refine Algebra.IsAlgebraic.of_ringHom_of_comp_eq (algebraMap R (⊥ : Subalgebra R A))
+    (RingHom.id A) (fun r ↦ ?_) Function.injective_id (by ext; rfl)
+  obtain ⟨y, hy⟩ := Algebra.mem_bot.1 r.2
+  use y
+  apply_fun algebraMap _ A using Subtype.val_injective
+  rw [← IsScalarTower.algebraMap_apply, hy]; rfl
```
Comment (alreadydone) [inline Mathlib/RingTheory/AlgebraicIndependent.lean:720]:
```suggestion
    (RingHom.id A) ?_ Function.injective_id (by ext; rfl)
  rintro ⟨_, r, rfl⟩; exact ⟨r, rfl⟩
```
and I think we should extract the lemma `Algebra.IsAlgebraic (↥⊥) A ↔ Algebra.IsAlgebraic R A`.

**Review by alreadydone — APPROVED** [review]

File: Mathlib/RingTheory/Localization/FieldRange.lean [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+variable {F A K L : Type*} [Field F] [CommRing A] [Algebra F A]
+  [Field K] [Algebra F K] [Algebra A K] [IsFractionRing A K] [Field L] [Algebra F L]
+
+section RingHom
+
+variable {g : A →+* L} {f : K →+* L}
+
+/-- If `A` is a ring with fraction field `K`, `g : A →+* L` lifts to `f : K →+* L`,
+then the image of `f` is the field genereated by the image of `g`. -/
```
[the remark this answers is not in the document]
Comment (acmepjz) [reply Mathlib/RingTheory/Localization/FieldRange.lean]: Done.

File: Mathlib/RingTheory/Localization/FieldRange.lean [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+theorem ringHom_fieldRange_eq_of_comp_eq
+    (h : RingHom.comp f (algebraMap A K) = g) :
+    f.fieldRange = Subfield.closure g.range := by
+  refine le_antisymm ?_ (by simp [← h, Set.range_comp_subset_range])
+  rintro x ⟨y, rfl⟩
+  obtain ⟨_, _, -, rfl⟩ := div_surjective (A := A) y
```
[the remark this answers is not in the document]
Comment (acmepjz) [reply Mathlib/RingTheory/Localization/FieldRange.lean:31]: Do you mean a `IsScalarTower` version?

File: Mathlib/RingTheory/AlgebraicIndependent.lean (hunk 1 above)
Comment (acmepjz) [reply Mathlib/RingTheory/AlgebraicIndependent.lean]: Done. The results are extracted into 'Mathlib/RingTheory/Algebraic.lean`. Note that some of the corresponding results for `IntermediateField` are not added yet; they should probably go to `FieldTheory/Adjoin.lean`.

File: Mathlib/RingTheory/Algebraic.lean [hunk 4]
```diff
@@ -337,6 +381,45 @@ theorem transcendental_algebraMap_iff {a : S} (h : Function.Injective (algebraMa
...
+    (RingHom.id S) (by rintro ⟨_, r, rfl⟩; exact ⟨r, rfl⟩) Function.injective_id (by ext; rfl)
+
+theorem isAlgebraic_bot_iff (h : Function.Injective (algebraMap R S)) {x : S} :
+    _root_.IsAlgebraic (⊥ : Subalgebra R S) x ↔ _root_.IsAlgebraic R x := by
+  refine (isAlgebraic_ringHom_iff_of_comp_eq (Algebra.botEquivOfInjective h) (RingHom.id S)
+    Function.injective_id ?_).symm
+  ext x
+  obtain ⟨y, rfl⟩ := (Algebra.botEquivOfInjective h).symm.surjective x
+  simp only [RingHom.coe_comp, RingHom.coe_coe, Function.comp_apply, AlgEquiv.apply_symm_apply]
+  rfl
```
Comment (alreadydone) [inline Mathlib/RingTheory/Algebraic.lean:402]:
```suggestion
  halg.of_ringHom_of_comp_eq (algebraMap R (⊥ : Subalgebra R S))
    (RingHom.id S) (by rintro ⟨_, r, rfl⟩; exact ⟨r, rfl⟩) Function.injective_id rfl

theorem isAlgebraic_bot_iff (h : Function.Injective (algebraMap R S)) {x : S} :
    _root_.IsAlgebraic (⊥ : Subalgebra R S) x ↔ _root_.IsAlgebraic R x :=
  isAlgebraic_ringHom_iff_of_comp_eq (Algebra.botEquivOfInjective h).symm (RingHom.id S)
    Function.injective_id (by rfl)
```

File: Mathlib/RingTheory/Algebraic.lean [hunk 5]
```diff
@@ -337,6 +381,45 @@ theorem transcendental_algebraMap_iff {a : S} (h : Function.Injective (algebraMa
...
+theorem algebra_isAlgebraic_bot_iff (h : Function.Injective (algebraMap R S)) :
+    Algebra.IsAlgebraic (⊥ : Subalgebra R S) S ↔ Algebra.IsAlgebraic R S := by
+  simp_rw [Algebra.isAlgebraic_def, isAlgebraic_bot_iff h]
+
+instance algebra_isAlgebraic_bot [Nontrivial R] : Algebra.IsAlgebraic R (⊥ : Subalgebra R S) := by
+  rw [Algebra.isAlgebraic_def]
+  intro x
+  rw [isAlgebraic_iff_isAlgebraic_val]
+  obtain ⟨y, hy⟩ := Algebra.mem_bot.1 x.2
+  exact hy ▸ isAlgebraic_algebraMap y
```
Comment (alreadydone) [inline Mathlib/RingTheory/Algebraic.lean:419]:
```suggestion
instance algebra_isAlgebraic_bot [Nontrivial R] : Algebra.IsAlgebraic R (⊥ : Subalgebra R S) :=
  ⟨by rintro ⟨_, x, rfl⟩; exact isAlgebraic_algebraMap _⟩
```

**Review by alreadydone — APPROVED** [review]
LGTM modulo #18984. The import increase is mainly due to AlgebraicIndependent importing FieldRange importing FieldTheory.Adjoin, but I think we're gonna import FieldTheory.Adjoin into AlgebraicIndependent sooner or later. Notice that AlgebraicIndependent is currently almost a leaf file, being transitively imported by only 2 other files.

**Comment by acmepjz** [comment]
> I think we're gonna import FieldTheory.Adjoin into AlgebraicIndependent sooner or later.

I think so, as most part of algebraic independent theory is about fields.

Maybe later we need to split AlgebraicIndependent into several files: subalgebra part, field part, and transcendence basis & transcendence degree part.

> Notice that AlgebraicIndependent is currently almost a leaf file, being transitively imported by only 2 other files.

Yeah, but once transcendence degree is defined, it should be used extensively in algebraic geometry. We also need separable transcendence basis & separably generated, etc. etc.

**Review by alreadydone — COMMENTED** [review]
Thanks :tada: 
maintainer merge

**Review by alreadydone — APPROVED** [review]

**Pull request closed without merging**